apitools/org.eclipse.pde.api.tools.ui/src/org/eclipse/pde/api/tools/ui/internal/completion/APIToolsJavadocCompletionProposalComputer.java

```diff
@@ -71,10 +71,10 @@
  */
 public class APIToolsJavadocCompletionProposalComputer implements IJavaCompletionProposalComputer {
 
-	private String fErrorMessage = null;
-	private Image fImageHandle = null;
-	private ASTParser fParser = null;
-	HashMap<String, Boolean> fExistingTags = null;
+	private volatile String fErrorMessage = null;
+	private volatile Image fImageHandle = null;
```

Comment:
If you want to MT harden this, it has to be an atomic reference, volatile is not enough, same for the other ones ...

Reply:
i harden it, it does not promise to make it totally mt safe. That would need much more work, as currently in the extreme case the sessionEnded() can even be called before collectExistingTags() because it is not waited on the backgroundthread after a timeout. However i have not observes such cases in real life. (logs an exception that fParser is null)